File: src/main/java/de/pflugradts/passbird/domain/model/transfer/Output.java
package de.pflugradts.passbird.domain.model.transfer;

import java.util.Objects;

/**
 * An Output represents data sent to the user through the
 * {@link de.pflugradts.passbird.application.UserInterfaceAdapterPort UserInterface}.
 */
public class Output {

    private final Bytes bytes;

    private Output(final Bytes bytes) {
        this.bytes = bytes;
    }

    public static Output of(final Bytes bytes) {
        return new Output(bytes);
    }

    public static Output empty() {
        return Output.of(Bytes.of());
    }

    public static Output of(final byte... bytes) {
        return Output.of(Bytes.of(bytes));
    }

    public Bytes getBytes() {
        return bytes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Output output = (Output) o;
        return Objects.equals(bytes, output.bytes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(bytes);
    }
}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/transfer/Output.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.emptyShell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.OutputFormatting.DEFAULT

class Output private constructor(val shell: Shell, val formatting: OutputFormatting?) {

    override fun equals(other: Any?): Boolean = when {
        (this === other) -> true
        (javaClass != other?.javaClass) -> false
        else -> shell == (other as Output).shell
    }
    override fun hashCode() = shell.hashCode()

    companion object {
        fun outputOf(shell: Shell, formatting: OutputFormatting = DEFAULT) = Output(shell, formatting)
        fun outputOf(byteArray: ByteArray, formatting: OutputFormatting = DEFAULT) = outputOf(shellOf(byteArray), formatting)
        fun emptyOutput() = outputOf(emptyShell())
    }
}

enum class OutputFormatting {
    DEFAULT,
    ERROR_MESSAGE,
    EVENT_HANDLED,
    HIGHLIGHT,
    NEST,
    OPERATION_ABORTED,
    SPECIAL,
}
